stats/redis.go

```diff
@@ -117,15 +136,86 @@ func (l *redisListener) HandleEvent(event events.Event) {
 	namespace := statsNamespace(key, event.Namespace())
 	bucket := event.BucketName()
 
-	var incr *redis.FloatCmd
-	_, err := l.client.Pipelined(func(pipe *redis.Pipeline) error {
-		incr = pipe.ZIncrBy(namespace, float64(numTokens), bucket)
-		pipe.ExpireAt(namespace, nearestHour())
-		return nil
-	})
+	l.queueStatsUpdate(namespace, numTokens, bucket)
+}
+
+// queueStatsUpdate queues a statsUpdate to be sent to redis via the batcher
+func (l *redisListener) queueStatsUpdate(namespace string, numTokens int64, bucket string) {
+	l.statsUpdatesLock.Lock()
+
+	l.pipe.ZIncrBy(namespace, float64(numTokens), bucket)
+	l.pipe.ExpireAt(namespace, nearestHour())
+
+	l.queuedUpdates++
+
+	l.statsUpdatesLock.Unlock()
+
+	select {
+	case l.notifyBatcher <- struct{}{}:
+		// Done
+	default:
+		// There's already a pending notification
+	}
+}
+
+// batcher aggregates stats updates. Any stats that are aggregated are guaranteed
+// to be sent within batchSubmitInterval, or once batchSize stats updates are queued,
+// whichever happens first.
+func (l *redisListener) batcher() {
+	timeout := time.After(l.batchDeadline)
+
+	for {
+		select {
+		case <-timeout:
+			// We've hit out deadline without submitting a batch, check if we have
+			// a batch to submit
+			l.statsUpdatesLock.Lock()
+			batchAvailable := l.queuedUpdates != 0
+			l.statsUpdatesLock.Unlock()
+
+			timeout = time.After(l.batchDeadline)
+
+			if !batchAvailable {
+				continue
+			}
+		case <-l.notifyBatcher:
+			// A new stats update has been submitted, check if we have enough to
+			// submit a batch
+			l.statsUpdatesLock.Lock()
+			batchReady := l.queuedUpdates >= l.batchSize
+			l.statsUpdatesLock.Unlock()
+
+			if !batchReady {
+				continue
+			}
+		}
+
+		l.statsUpdatesLock.Lock()
+
+		pipe := l.pipe
+		l.pipe = l.client.Pipeline()
+		l.queuedUpdates = 0
+
+		l.statsUpdatesLock.Unlock()
+
+		go l.submitBatch(pipe)
+
+		timeout = time.After(l.batchDeadline)
+	}
+}
+
+// submitBatch executes the provided pipeline and checks for errors.
+func (l *redisListener) submitBatch(pipe *redis.Pipeline) {
+	cmds, err := pipe.Exec()
+	if err != nil {
+		logging.Printf("RedisStatsListener.HandleEvent pipeline error %v", err)
+	}
+
+	for _, cmd := range cmds {
+		if cmd.Err() == nil {
+			continue
+		}
 
-	if err != nil || incr.Err() != nil {
-		logging.Printf("RedisStatsListener.HandleEvent error (%s, %s, %d) %v, %v",
-			namespace, bucket, numTokens, err, incr.Err())
+		logging.Printf("RedisStatsListener.HandleEvent error (%s) %v", cmd.String(), cmd.Err())
```

Comment:
Let's move this under `if cmd.Err() != nil` and get rid of the continue. That conforms to https://github.com/golang/go/wiki/CodeReviewComments#indent-error-flow

Reply:
Will do, I have no idea why I did that flow